modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js

```diff
@@ -0,0 +1,482 @@
+
+import GL from '@luma.gl/constants';
+import {
+  boundsContain,
+  getTriangleVertices,
+  parseData,
+  scaleToAspectRatio,
+  scaleTextureCoordiantes
+} from './heatmap-layer-utils';
+import {Buffer, Transform, getParameter, isWebGL2} from '@luma.gl/core';
+import {CompositeLayer, log} from '@deck.gl/core';
+import TriangleLayer from './triangle-layer';
+import {getFloatTexture} from '../utils/resource-utils';
+import {defaultColorRange, colorRangeToFlatArray} from '../utils/color-utils';
+import weights_vs from './weights-vs.glsl';
+import weights_fs from './weights-fs.glsl';
+import vs_max from './max-vs.glsl';
+
+const RESOLUTION = 2; // (number of common space pixels) / (number texels)
+const SIZE_2K = 2048;
+
+const defaultProps = {
+  getPosition: {type: 'accessor', value: x => x.position},
+  getWeight: {type: 'accessor', value: x => 1},
+  intensity: {type: 'number', min: 0, max: 1, value: 1},
+  radiusPixels: {type: 'number', min: 1, max: 100, value: 30},
+  colorRange: defaultColorRange,
+  enhanceFactor: {type: 'number', min: 1, max: 100, value: 20}
+};
+
+export default class HeatmapLayer extends CompositeLayer {
+  initializeState() {
+    const {gl} = this.context;
+    if (!isWebGL2(gl)) {
+      log.error('HeatmapLayer is not supported on this browser, requires WebGL2')();
+    }
+    const textureSize = Math.min(SIZE_2K, getParameter(gl, gl.MAX_TEXTURE_SIZE));
+    const weightsTexture = getFloatTexture(gl, {
+      width: textureSize,
+      height: textureSize,
+      parameters: {
+        [GL.TEXTURE_MAG_FILTER]: GL.LINEAR,
+        [GL.TEXTURE_MIN_FILTER]: GL.LINEAR
+      }
+    });
+    const maxWeightsTexture = getFloatTexture(gl); // 1 X 1 texture
+    this.state = {
+      textureSize,
+      weightsTexture,
+      maxWeightsTexture,
+      weightsTransform: new Transform(gl, {
+        id: `${this.id}-weights-transform`,
+        vs: weights_vs,
+        _fs: weights_fs,
+        modules: ['project32'],
+        elementCount: 1,
+        _targetTexture: weightsTexture,
+        _targetTextureVarying: 'weightsTexture'
+      }),
+      maxWeightTransform: new Transform(gl, {
+        id: `${this.id}-max-weights-transform`,
+        _sourceTextures: {
+          inTexture: weightsTexture
+        },
+        _targetTexture: maxWeightsTexture,
+        _targetTextureVarying: 'outTexture',
+        vs: vs_max,
+        elementCount: textureSize * textureSize
+      }),
+      zoom: null,
+      triPositionBuffer: new Buffer(gl, {
+        data: getTriangleVertices({xMin: -1, yMin: -1, xMax: 1, yMax: 1, addZ: true}),
+        accessor: {size: 3}
+      }),
+      triTexCoordBuffer: new Buffer(gl, {
+        data: getTriangleVertices(),
+        accessor: {size: 2}
+      })
+    };
+  }
+
+  shouldUpdateState({changeFlags}) {
+    // Need to be updated when viewports
+    return changeFlags.somethingChanged;
+  }
+
+  // returns visible world bounds [[minLong, minLat], [maxLong, maxLat]
+  getVisibleWorldBounds() {
```

Comment:
Use underscore prefix for private methods

Reply:
Done.